Refactor Centralize session and default description

Belal-Alkhatib opened the pull request:
# Pull Request Template

## Description
This commit introduces a new annotation, `@RequiresSessionId`, to handle session ID injection in API requests and refactors the codebase to remove explicit session ID handling from repositories and data sources, relying instead on an interceptor for session management.
---

## Changes Made
List the changes introduced in this pull request:

- feat: add RequiresSessionId annotation to handle session ID injection  
- RetrofitClient updated to accept AuthenticationLocalDataSource and CryptoData, check for "X-Require-Session" header, and inject session ID as query parameter  
- MovieApiService methods updated to use the new annotation and remove explicit session_id query parameter  
- apiServiceModule updated to provide AuthenticationLocalDataSource and CryptoData to RetrofitClient  
- Refactor: Removed session ID dependency from MovieRepositoryImpl and MovieRemoteDataSourceImpl  
- Refactor: Removed session ID from ProfileApiService and related repository and data source calls, added header annotation for session requirement  
- Refactor: Removed session ID from TV show API calls (getTvShowDetailsById, getRatedTvShows, setTvShowRate, deleteTvShowRate)  
- Refactor: Changed movieId parameter type from Int to Long in addMovieToList and related functions  
- Removed redundant sessionId and description parameters from UserListRemoteSource and AuthenticationRepository  
- Test: Refactored repository tests to remove redundant sessionId and cryptoData, simplify setup, and standardize naming

---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
This refactor centralizes session ID management within the networking layer, reduces redundancy, and improves code maintainability.

Belal-Alkhatib committed:
refactor: remove session_id and description from user list API calls

Belal-Alkhatib committed:
refactor: use Long for movie ID in "add to list" functionality

Belal-Alkhatib committed:
feature: add RequiresSessionId annotation to handle session ID injection

Belal-Alkhatib committed:
refactor: Remove session ID dependency from movie repository and remote data source

Belal-Alkhatib committed:
refactor: remove session ID from profile API and repository calls

Belal-Alkhatib committed:
refactor: remove session ID from TV show API calls

Belal-Alkhatib committed:
refactor: update parameter type and remove redundant parameters

Belal-Alkhatib committed:
test: refactor repository tests

Belal-Alkhatib committed:
Merge branch 'refs/heads/develop' into refactor/centralize-default-desc-and-session-id

# Conflicts:
#	domain/src/main/java/com/amsterdam/domain/repository/UserListRepository.kt
#	repository/src/main/java/com/amsterdam/repository/datasource/remote/UserListRemoteSource.kt
#	repository/src/main/java/com/amsterdam/repository/repository/UserListRepositoryImpl.kt
#	repository/src/test/java/com/amsterdam/repository/repository/UserListRepositoryImplTest.kt

Belal-Alkhatib committed:
refactor: remove redundant UserListRepository methods and parameters

mohamedd-hassan commented on remoteDatasource/src/main/java/com/amsterdam/remotedatasource/utils/RequiresSessionId.kt:
```diff
@@ -0,0 +1,5 @@
+
+@Retention(AnnotationRetention.RUNTIME)
+@Target(AnnotationTarget.FUNCTION)
+annotation class RequiresSessionId
```
you introduced this annotation but it was never used

you can do this

```
            val invocation = originalRequest.tag(Invocation::class.java)

            val requireSession = invocation?.method()
                ?.getAnnotation(RequiresSessionId::class.java) != null

            if (requireSession) {
                val sessionId = runBlocking {
                    cryptoData.decryptString(authenticationLocalDataSource.getCachedSessionId())
                }

                if (!sessionId.isNullOrBlank()) {
                    originalHttpUrlBuilder.addQueryParameter(SESSION_PARAM_NAME, sessionId)
                }
            }

            val newRequest = originalRequest.newBuilder()
                .url(originalHttpUrlBuilder.build())
                .header(TOKEN_HEADER_NAME, "Bearer $token")
                .build()
```
in the retrofit client

so your api service would just be

```
    @RequiresSessionId
    @FormUrlEncoded
    @POST("movie/{movie_id}/rating")
    suspend fun postMovieRating(
        @Path("movie_id") movieId: Long,
        @Field("value") rate: Float,
    ): RatingResponse
```

to make it even safer if you want to use this annotation

Belal-Alkhatib replied:
i already made this😅 @mohamedd-hassan, could you take another look ?

mohamedd-hassan replied:
you created the new annotation but its not used anywhere and you use the header directly so this is just a suggestion if you want to be able to use it

Belal-Alkhatib replied:
Yes, you’re right. I made the solution you mentioned, you can see it now.

Belal-Alkhatib committed:
refactor: simplify MovieDetailsViewModel tests

This commit refactors the `MovieDetailsViewModelTest` class to:

- Utilize JUnit 5 extensions for better test structure.
- Employ Turbine for testing Flow emissions.
- Improve test clarity and conciseness by:
    - Reducing boilerplate code for ViewModel initialization.
    - Focusing each test on a single behavior.
    - Using descriptive test names.
- Group related tests using regions.
- Define shared test data as private properties.

Belal-Alkhatib committed:
Merge branch 'develop' into refactor/centralize-default-desc-and-session-id

# Conflicts:
#	domain/src/main/java/com/amsterdam/domain/repository/TvShowRepository.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImpl.kt
#	remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImpl.kt
#	repository/src/main/java/com/amsterdam/repository/datasource/remote/TvShowsRemoteSource.kt
#	repository/src/main/java/com/amsterdam/repository/repository/MovieRepositoryImpl.kt

Belal-Alkhatib committed:
refactor: update function signatures and remove duplicate calls
